### ISSUE TITLE
Rule proposal: Only allow `Array#reduce()` for summing/multiplying numbers

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#318]
`Array#reduce()` can often make code less readable and sometimes also [slower](https://twitter.com/sophiebits/status/1099059115446292481). A `for-of` loop is usually a better choice.

See this Twitter thread: https://twitter.com/sophiebits/status/1099014182261776384?s=21

Not sure how strict we should/can be, but my thinking is that we only allow `Array#reduce()` when the callback returns a number and the `initialValue` is a number if set.

**Comment by bradzacher** [comment]
I don't necessarily agree with this.

The reduce function is a boundary. Mutating the accumulator within that boundary is fine from a purity standpoint, as long as the initialValue isn't a variable reference.

```TS
arr.reduce((acc, value) => {
    acc[calcdKey] = value;
    return acc;
}, {});
```

The thing I'd lint would be:
Ensuring that the initial value is not a variable reference (unless the variable's type is `number | string | boolean`).

**Comment by sindresorhus** [comment]
The intention is not about purity, but rather readability. A lot of people find `.reduce` a lot less readable than a for-loop.

**Comment by sindresorhus** [comment]
> The thing I'd lint would be:
Ensuring that the initial value is not a variable reference (unless the variable's type is number | string | boolean).

I agree that would be useful too, but I'd rather see that as a separate rule.

**Comment by JoshuaKGoldberg** [comment]
This rule has sat here for a little over three years with only a couple of votes in favor. It's also not entirely TypeScript-specific. I'm going to go ahead and close this for housekeeping purposes.

Feel free to post back here if there is a compelling TypeScript-specific use case not yet discussed!

My recommendation to anybody particularly interested in implementing this rule would be to make it in a standalone repository and sharing it around to build community buy-in. https://eslint.org/docs/developer-guide/working-with-rules

**Comment by sindresorhus** [comment]
For anyone looking, this was added here: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-reduce.md
